### PR TITLE
Optimize experimental Kafka scaler and fix consumer group logic

### PR DESCRIPTION
**Pull request by adrien-f**
TL;DR:

* Only request metadata on the topics needed
* Fix the logic to fetch topics from the consumer groups

---

I know we decided to implement the IAM auth in the sarama based scaler (the original) but there is still an optimization that would serve current users of the experimental scaler and potentially lower their Keda load and their broker.

Firstly, the Metadata request is not scoped for the requested topics (which can be empty):

https://github.com/kedacore/keda/blob/bcaf5c07e785e3e58e3be4e3707b518fdef6acde/pkg/scalers/apache_kafka_scaler.go#L425-L427

If the list of topics is empty, we enter the branch to detect the topics & permissions based on the consumer group activity here:

https://github.com/kedacore/keda/blob/bcaf5c07e785e3e58e3be4e3707b518fdef6acde/pkg/scalers/apache_kafka_scaler.go#L441C3-L441C14

On my AWS MSK cluster, the version for the response is not supported by the segment-io library which causes the following to error out, as seen in https://github.com/segmentio/kafka-go/issues/1212 (probably):

```
Got non-zero number of bytes remaining: 10
```

Alright, for the case of debugging, let's ignore the error. What happens next? From what I can see, this variable `describeGrp` is unused, and because the MetadataRequest is empty, the whole list of topics is processed and returned with their partitions. All the time. Which then cause Keda to request all consumer & producer offsets again and again, etc...

So to me, currently the behavior is buggy. The `describeGrp` should be processed to extract the list of topics and partitions for the consumer group which I also correct in this PR.

### Checklist

- [ ] When introducing a new scaler, I agree with the [scaling governance policy](https://github.com/kedacore/governance/blob/main/SCALERS.md)
- [ ] I have verified that my change is according to the [deprecations & breaking changes policy](https://github.com/kedacore/governance/blob/main/DEPRECATIONS.md)
- [ ] Tests have been added
- [ ] Changelog has been updated and is aligned with our [changelog requirements](https://github.com/kedacore/keda/blob/main/CONTRIBUTING.md#Changelog)
- [ ] A PR is opened to update our Helm chart ([repo](https://github.com/kedacore/charts)) *(if applicable, ie. when deployment manifests are modified)*
- [ ] A PR is opened to update the documentation on ([repo](https://github.com/kedacore/keda-docs)) *(if applicable)*
- [ ] Commits are signed with Developer Certificate of Origin (DCO - [learn more](https://github.com/kedacore/keda/blob/main/CONTRIBUTING.md#developer-certificate-of-origin-signing-your-work))

Fixes #

Relates to #5531

### CODE REVIEW

File: pkg/scalers/apache_kafka_scaler.go [hunk 1]
```diff
@@ -422,58 +422,77 @@ func getApacheKafkaClient(ctx context.Context, metadata apacheKafkaMetadata, log
...
+
+func (s *apacheKafkaScaler) getTopicPartitionsFromConsumerGroup(ctx context.Context) (map[string][]int, error) {
+	describeGrp, err := s.client.DescribeGroups(ctx, &kafka.DescribeGroupsRequest{
+		Addr: s.client.Addr,
+		GroupIDs: []string{
+			s.metadata.group,
+		},
+	})
+
+	// Currently, the request could fail because of an unsupported version of the protocol
```
Comment (dttung2905) [inline pkg/scalers/apache_kafka_scaler.go:463]: If this is a known error, can we put a link to the issue in the comment so that we can easily trace it back in the future ?

**Review by dttung2905 — COMMENTED** [review]
Hi @adrien-f ,
Thanks for refactoring this piece of code. Generally its LGTM. Just a couple of small comments for me
Since there is a new helper function `getTopicPartitionsFromConsumerGroup` created, would that be possible to create a unit test for it ?

File: pkg/scalers/apache_kafka_scaler.go (hunk 1 above)
Comment (adrien-f) [reply pkg/scalers/apache_kafka_scaler.go:463]: Done 👍

**Comment by adrien-f** [comment]
@dttung2905 thanks you for the review! I will look into adding unit tests, the issue being that this lib does not have an interface to create mock on. I will investigate and see what's possible to do.

**Comment by zroubalik** [comment]
any update on this, please?
